From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Thu, 12 Sep 2024 11:30:23 +0200
Subject: [PATCH] Regression: Fix MiniPlaceholders

Fixed by adding a ResourceLocation(String id) constructor back.

diff --git a/src/main/java/net/minecraft/resources/ResourceLocation.java b/src/main/java/net/minecraft/resources/ResourceLocation.java
index 1967c43ee3a12e63365cc40ee6565307e2fd73cf..40d0e3c67aff9cb073f4a88c00895b42e72eca56 100644
--- a/src/main/java/net/minecraft/resources/ResourceLocation.java
+++ b/src/main/java/net/minecraft/resources/ResourceLocation.java
@@ -292,4 +292,27 @@ public final class ResourceLocation implements Comparable<ResourceLocation> {
             return new JsonPrimitive(resourceLocation.toString());
         }
     }
+
+    // Abomination start - fix MiniPlaceholder regression
+    public ResourceLocation(String id) {
+        this(decompose(id, ':'));
+    }
+
+    private ResourceLocation(String[] id) {
+        this(id[0], id[1]);
+    }
+
+    protected static String[] decompose(String id, char delimiter) {
+        String[] strings = new String[]{"minecraft", id};
+        int i = id.indexOf(delimiter);
+        if (i >= 0) {
+            strings[1] = id.substring(i + 1);
+            if (i >= 1) {
+                strings[0] = id.substring(0, i);
+            }
+        }
+
+        return strings;
+    }
+    // Abomination end
 }
