From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Sun, 15 Sep 2024 17:53:14 +0200
Subject: [PATCH] linearmanager


diff --git a/src/main/java/abomination/LinearManager.java b/src/main/java/abomination/LinearManager.java
new file mode 100644
index 0000000000000000000000000000000000000000..eb5d0c3ce292255468bb558d41614b7c6e715bf7
--- /dev/null
+++ b/src/main/java/abomination/LinearManager.java
@@ -0,0 +1,32 @@
+package abomination;
+
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
+
+public class LinearManager extends Thread {
+    private static final org.slf4j.Logger LOGGER = com.mojang.logging.LogUtils.getLogger();
+
+    Thread[] workers = null;
+    private boolean close = false;
+
+    public static LinearManager instance = new LinearManager();
+
+    public final ObjectArrayList<LinearRegionFile> regionCache = new ObjectArrayList();
+
+    public void begin() {
+        
+    }
+
+    public void addRegion(LinearRegionFile region) {
+        regionCache.add(region);
+    }
+
+    public void run() {
+        while (!close) {
+
+            
+
+            try { Thread.sleep(100); } catch(InterruptedException ignored) {}
+        }
+    }
+
+}
diff --git a/src/main/java/abomination/LinearRegionFile.java b/src/main/java/abomination/LinearRegionFile.java
index 92bb140c4cb290b1872b745c8101097fec18f644..68debbd0cdd2d8a6899f8dd768564440c6f15c7a 100644
--- a/src/main/java/abomination/LinearRegionFile.java
+++ b/src/main/java/abomination/LinearRegionFile.java
@@ -143,6 +143,7 @@ public class LinearRegionFile extends Thread {
             }
 
             this.start();
+            LinearManager.instance.addRegion(this);
         } catch (IOException e) {
             throw new RuntimeException("Failed to open region file " + this.regionFile, e);
         }
