From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Sun, 15 Sep 2024 17:53:21 +0200
Subject: [PATCH] Revert "linearmanager"

This reverts commit b9121a750a254166119f0f35fd67bcabf09f1f47.

diff --git a/src/main/java/abomination/LinearManager.java b/src/main/java/abomination/LinearManager.java
deleted file mode 100644
index eb5d0c3ce292255468bb558d41614b7c6e715bf7..0000000000000000000000000000000000000000
--- a/src/main/java/abomination/LinearManager.java
+++ /dev/null
@@ -1,32 +0,0 @@
-package abomination;
-
-import it.unimi.dsi.fastutil.objects.ObjectArrayList;
-
-public class LinearManager extends Thread {
-    private static final org.slf4j.Logger LOGGER = com.mojang.logging.LogUtils.getLogger();
-
-    Thread[] workers = null;
-    private boolean close = false;
-
-    public static LinearManager instance = new LinearManager();
-
-    public final ObjectArrayList<LinearRegionFile> regionCache = new ObjectArrayList();
-
-    public void begin() {
-        
-    }
-
-    public void addRegion(LinearRegionFile region) {
-        regionCache.add(region);
-    }
-
-    public void run() {
-        while (!close) {
-
-            
-
-            try { Thread.sleep(100); } catch(InterruptedException ignored) {}
-        }
-    }
-
-}
diff --git a/src/main/java/abomination/LinearRegionFile.java b/src/main/java/abomination/LinearRegionFile.java
index 68debbd0cdd2d8a6899f8dd768564440c6f15c7a..92bb140c4cb290b1872b745c8101097fec18f644 100644
--- a/src/main/java/abomination/LinearRegionFile.java
+++ b/src/main/java/abomination/LinearRegionFile.java
@@ -143,7 +143,6 @@ public class LinearRegionFile extends Thread {
             }
 
             this.start();
-            LinearManager.instance.addRegion(this);
         } catch (IOException e) {
             throw new RuntimeException("Failed to open region file " + this.regionFile, e);
         }
