From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Mon, 16 Sep 2024 04:50:15 +0200
Subject: [PATCH] linear


diff --git a/src/main/java/abomination/LinearManager.java b/src/main/java/abomination/LinearManager.java
index 70c5953a3a6905d42ed0963de6fa3fc6d7a28ad0..0136bf342ffc3ead54c928735a180a3a0d5df392 100644
--- a/src/main/java/abomination/LinearManager.java
+++ b/src/main/java/abomination/LinearManager.java
@@ -43,11 +43,13 @@ public class LinearManager extends Thread {
 
     public void addRegion(LinearRegionFile region) {
         synchronized(regionCache) {
+            addedRegions++;
             regionCache.add(region);
         }
     }
 
     long lastReport = 0;
+    int addedRegions = 0;
     int toFlushCount = 0;
     int removedCount = 0;
 
@@ -56,8 +58,13 @@ public class LinearManager extends Thread {
             synchronized(regionCache) {
                 long timestamp = System.currentTimeMillis();
 
-                if (timestamp - lastReport > 60000 && abomination.config.UnsupportedWipOptionsWillChange.linearVerbose)
-                    LOGGER.info("regionCache: " + regionCache.size() + " toFlush: " + toFlush.size() + " toFlushCount: " + toFlushCount + " removedCount: " + removedCount);
+                if (timestamp - lastReport > 60000 && abomination.config.UnsupportedWipOptionsWillChange.linearVerbose) {
+                    LOGGER.info("regionCache: " + regionCache.size() + " toFlush: " + toFlush.size() + " addedRegions: " + addedRegions + " toFlushCount: " + toFlushCount + " removedCount: " + removedCount);
+                    addedRegions = 0;
+                    toFlushCount = 0;
+                    removedCount = 0;
+                    lastReport = timestamp;
+                }
 
                 for (int i = 0 ; i < regionCache.size() ; i++) {
                     LinearRegionFile file = regionCache.get(i);
